src/sentry/api/endpoints/organization_events.py

```diff
@@ -218,64 +234,155 @@ def get(self, request, organization):
+            return Response({
+                'detail': 'Invalid query.'  # TODO(lb): is this ok?
+            }, status=400)
+
+        if lookup_keys:
+            try:
+                top_values = self.get_top_values(keys=lookup_keys, **snuba_args)
+            except (KeyError, SnubaError):
+                logger.info(  # TODO(lb): should this be level error?
+                    'api.organization-events-heatmap',
+                    extra={
+                        'organization_id': organization.id,
+                        'user_id': request.user.id,
+                        'keys': keys,
+                        'snuba_args': snuba_args,
+                    }
+                )
+                return Response({
+                    'detail': 'Invalid query.'  # TODO(lb): is this ok?
+                }, status=400)
+        else:
+            top_values = []
 
         if non_tag_lookup_keys:
-            tag_keys.update(self.handle_non_tag_keys(non_tag_lookup_keys, snuba_args))
+            self.handle_non_tag_keys(non_tag_lookup_keys, snuba_args, top_values)
+
+        tag_keys = self._create_tag_objects(keys, total_count, top_values)
+
+        # sort heatmap categories in the same order requested.
+        tag_keys = sorted(tag_keys, key=lambda x: keys.index(x.key))
 
         return Response(serialize(tag_keys, request.user))
 
-    def handle_non_tag_keys(self, keys, snuba_args):
-        result = set([])
+    def handle_non_tag_keys(self, keys, snuba_args, top_values):
         for key in keys:
+
+            data = self._query_non_tag_data(self.NON_TAG_KEYS[key], snuba_args)
+
             if key == 'project.name':
-                data = self._query_non_tag_data('project_id', snuba_args)
                 projects = Project.objects.filter(id__in=snuba_args['filter_keys']['project_id'])
-                for project_data in data:
-                    project = projects.filter(id=project_data['project_id'])[0]
-                    project_data['key'] = 'project.name'
-                    project_data['value'] = project.slug
-            result.add(self._create_tag_key_tag_value_objects('project', data))
-        return result
+                for value in data:
+                    project_slug = projects.filter(id=value['project_id'])[0].slug
+                    value['tags_value'] = project_slug
+                    value['tags_key'] = key
+            else:
+                for value in data:
+                    value['tags_key'] = key
+                    value['tags_value'] = value[self.NON_TAG_KEYS[key]]
+
+            top_values += data
+
+        # order combined values by count
+        top_values = sorted(top_values, key=lambda x: x['count'])
 
     def _query_non_tag_data(self, key, snuba_args):
+        snuba_args = deepcopy(snuba_args)
         data = raw_query(
             groupby=[key],
-            aggregations=snuba_args.get('aggregations', []) + [
+            conditions=snuba_args.pop('conditions', []) + [
```

Comment:
We should limit the list of non tag values in line with tag values (top 9 or 10 values are returned I believe?)

Reply:
https://github.com/getsentry/sentry/blob/master/src/sentry/tagstore/base.py#L22